### ISSUE TITLE
Support Substack

### ISSUE THREAD

**Issue by gustavo-depaula** [Maluen/code-medium#11]
The extension is great for doing code blocks in Medium. Any thoughts on supporting [Substack](substack.com), or do you know a similar solution to substack?

**Comment by Maluen** [comment]
No plans on supporting Substack for now, as this extension was designed and built with medium only in mind. I've never used Substack myself so first we would indeed have to look at the current solutions and at the editor itself.

A main requirement is for them to allow gist embedding. From a quick search looks like it isn't possible?

**Comment by gustavo-depaula** [comment]
@Maluen it is possible:
<img width="713" alt="image" src="https://user-images.githubusercontent.com/732422/158061495-37596f8a-3194-47c6-a899-d85130c956ad.png">

But IDK why this text about bidirectional unicode text happens

**Comment by Maluen** [comment]
The unicode warning is specific to gist, see https://github.blog/changelog/2021-10-31-warning-about-bidirectional-unicode-text/ I suggest trying with other gists, e.g. by creating one from scratch.

Substack supporting gist is a good first step as we can reuse all the background part of the extension and also all the UI for creating/updating gists/logging-in.

So the part still left to implement would be extending their WYSIWYG editor:
- button to show the "create new gist UI"
- double click on existing gists to edit them
- CSS to make sure the extension UI doesn't break their editor.

**Comment by gustavo-depaula** [comment]
Hey! I was able to create a new service to show the button:
<img width="673" alt="image" src="https://user-images.githubusercontent.com/732422/158078496-d27c7198-8f55-4f26-acd6-50cb5c77eab3.png">

But for some reason the event listeners don't seem to work inside the dropdown... Even when using `window.addEventListener("click", console.log)`, clicks on the dropdown items are not triggering clicks. Maybe there are some `event.stopPropagation/preventDefault` that are "swallowing" these events? 

Here's how I'm creating the button: https://github.com/gustavo-depaula/code-medium/commit/bf3a063c60b83d03f97e2e1b81cbc06bb733afcf#diff-32bbb3346e15f8054cffefe0a61d4a814a64dcf6d61674aeda6795c56ce21c6bR68-R82

**Comment by Maluen** [comment]
>Hey! I was able to create a new service to show the button:

Nice work! Looks promising! I think in terms of CSS we'll have to separate the medium styles from the substack styles to limit regressions. E.g. by using a different CSS file for each (with some common imported styles if needed, better if not).

>Maybe there are some event.stopPropagation/preventDefault that are "swallowing" these events?

I'll have to check on the actual website. The last resort is `window.addEventListener("click", handler, true)`  (notice the third argument being `true`), where the `handler` function checks if `event.target` is the button or an element inside the button.

You can first try passing `true` when attaching the event to the element itself `buttonEl.addEventListener('click', this.handleCreateGistClick, true);`

**Comment by gustavo-depaula** [comment]
>  I think in terms of CSS we'll have to separate the medium styles from the substack styles to limit regressions.

Absolutely! I'm just trying to get it to work before doing all these design/architectural decisions.

> You can first try passing true when attaching the event to the element itself

It didn't work... Might it perhaps be because it's not `button`? 
<img width="880" alt="image" src="https://user-images.githubusercontent.com/732422/158177806-51729a6a-dd09-4799-b657-502fb3fb1eda.png">

**Comment by Maluen** [comment]
I've checked the substack editor. The dropdown closes on mousedown, thus the click never registers. Code below works:

```javascript
buttonEl.addEventListener('mousedown', this.handleCreateGistClick);
```

The `handleCreateGistClick` needs to be adapted for substack as with other functions.

**Comment by Maluen** [comment]
See https://github.com/Maluen/code-medium/pull/12 and https://github.com/Maluen/code-medium/pull/12/commits/465f7c2c9135e08392dd5df47fea3627be34c45c

I've also changed `insertGistIntoPost` to simulate a `paste` event with the gist URL.

For gist editing, I've noticed that they are not embedding the IFRAME with the gist, they are actually pulling the HTML and adding it inside a regular DIV. This means a different method needs to be used to update the gist.

**Comment by Maluen** [comment]
https://github.com/Maluen/code-medium/pull/12/commits/3e105897929aecb69f3b4caae039e5520cff5061

- Added `dblclick` handler to trigger edit UI when double clicking on gists.
- Added code to delete a gist (`deleteGistIntoPost')

The main thing still left to do in SubstackService is the `updateGistIntoPost` function, which is supposed to refresh a gist with its new version after the editing is finished. For now I've tried deleting and then repasting the gist URL, but for some reasons the editor reads it as a plain URL and doesn't convert it into an embed. Will need to find a workaround.

![image](https://user-images.githubusercontent.com/1434239/158269640-0a67ca35-0804-4472-9dad-9781336ee541.png)

**Comment by gustavo-depaula** [comment]
Nice!! Wouldn't have found the event stuff. Thanks

I think it might be a problem with the link itself. Pasting manually:
```
https://gist.github.com/gustavo-depaula/aeeccfa7056acf75836b57a23c063c89  -> this creates an embed
https://gist.github.com/aeeccfa7056acf75836b57a23c063c89 -> this doesn't
```
(the difference is the username, that must be simple to inject, gonna try it)

**Comment by gustavo-depaula** [comment]
done the username injection in https://github.com/gustavo-depaula/code-medium/commit/d46cd3f49d0d22e023d80309acf6d243fd7378c8

insertion is now working as expected:
<img width="804" alt="image" src="https://user-images.githubusercontent.com/732422/158287767-69369d04-07fb-45bb-8408-81b70d7c3029.png">

**Comment by gustavo-depaula** [comment]
Here's what I'm thinking about update: we could trigger a delete/backspace, that would delete the current embed, and then call the insert function again. That would fetch the new content

**Comment by gustavo-depaula** [comment]
Just noticed that there's already a `simulateBackspaceKeydown` function and that `deleteGistIntoPost` uses it.
Got the update working in by using the delete/insert-again approach: https://github.com/gustavo-depaula/code-medium/commit/3abf81181848b2fb481675d8028834a2769f752b


https://user-images.githubusercontent.com/732422/158298514-694cfd9b-9273-4bbf-b6b7-98c238500f5a.mp4


Still being tormented by the bidirectional unicode text thingy, though...

**Comment by gustavo-depaula** [comment]
Thought in just deleting the warning by doing:
```js
      const bidirectionalUnicodeTextWarn = document.querySelector('.flash-warn');
      bidirectionalUnicodeTextWarn.parentElement.removeChild(bidirectionalUnicodeTextWarn);
```
but (obviously) doesn't work because when it'll be published the warning will appear hahahaha

**Comment by Maluen** [comment]
Yes, the update wasn't working in my tests because of the different URL structure. Nice find on that!

About the unicode warning, that is indeed very annoying, I don't think we should add Substack support to the extension if every gist shows the warning.

I tried creating a new one-line gist directly from gist.github.com and then pasting it into Substack, it still shows the warning, which makes me think it might be related to how Substack is fetching the gist on their server-side. Maybe a bug on their end?

**Comment by gustavo-depaula** [comment]
Yes... It's my best theory also. Using the network request, we're able to see that the message is returned on the response of their gist embed API.

Thus, I have contacted substack support reporting this bug. Let's see what they say.

Aside from that, is there any feature/behavior left to implement?

**Comment by Maluen** [comment]
>Thus, I have contacted substack support reporting this bug. Let's see what they say.

Cool, let's hope for a quick solution.

>Aside from that, is there any feature/behavior left to implement?

Not much. There is code in SubstackService left from the MediumService to prevent interaction with the editor while the extension UI is visible (i.e. while gists are being created/edited): https://github.com/gustavo-depaula/code-medium/blob/3abf81181848b2fb481675d8028834a2769f752b/src/content/services/SubstackService.js#L72-L102

That is needed when the extension UI is displayed on the right side in big screens. In that case the user shouldn't be able to edit the article, but they should still be able to select text if possible. However even just allowing text selection could break the extension when clicking on "Create gists" since the focus could have been lost, the cursor have changed position, etc.

Other than that there is the CSS separation I talked before, where content.scss can be splitted into medium.scss and substack.scss.

I think I will also want to rename `content_iframe.js` into `medium_iframe.js`.

Basically making it clear what is medium and what is substack.

**Comment by Maluen** [comment]
Code restructured here: https://github.com/Maluen/code-medium/pull/12/commits/0632335bbaa201e06b05771818ca8252ddd88bf2

**Comment by gustavo-depaula** [comment]
hey, closing this as I'm not using substack and never got an answer

thanks for your support, though! great extension

**Comment by KyMidd** [comment]
Did we ever receive an update from Substack on the unicode warnings? I'm also pushing for an answer, as I use gist extensively embedded in Medium pages, and as I'm moving to Substack I'm seeing all these warnings. 

Thanks for all ya'll do.

**Comment by Maluen** [comment]
I'm still seeing unicode warnings for most gists on substack.

I just tried to create a new gist (directly from github) with only the word "test" inside, no spaces, no newlines, and substack is still showing the unicode warnings when embedding.

The same warning also shows on the published post, so it isn't just a draft/preview issue.

**Comment by KyMidd** [comment]
I don't work at github, but know some folks well who do. I'm escalating this with GitHub support. 

I see the error message coming from github directly, [here's a link to a gist](https://gist.github.com/KyMidd/fb0cd10efd25909a66e7ff2016327d69.js) with string "asdfasd" only, no returns or characters, not copied from anywhere, no invisible unicode characters, and it still shows the error. 

I'll let you know when/if we get anywhere :) Thanks @Maluen!

**Comment by Maluen** [comment]
Indeed, Substack isn't embedding the gist as an iframe like Medium, it is pullilng the HTML itself from the gist URL, which contains the warning as visible in your link.

**Comment by KyMidd** [comment]
Oh, so this isn't a GitHub change. That's a bummer, I don't have any sway at Substack. I'm not very hopeful to evince changes over there. Okay, time to migrate all my (hundreds) of gists to direct embedded codeblocks. Thanks Maluen, I appreciate ya.

**Comment by Maluen** [comment]
Looking at [your link](https://gist.github.com/KyMidd/fb0cd10efd25909a66e7ff2016327d69.js) and executing the JS (it's a bunch of document.write calls), we can see that the unicode warnings are actually inside `template` elements and aren't actually rendered.

![image](https://github.com/Maluen/code-medium/assets/1434239/45f0d1a7-a1d8-4651-a3bb-8b71e797170d)

The rendered page shows no warning

![image](https://github.com/Maluen/code-medium/assets/1434239/fd8957a8-cd9a-4dda-b546-3c2212a60366)

However the warnings are shown when embedded in Substack.

Looking at substack. they are doing an API call to `api/v1/github/gist` that returns the following JSON:

```json
{
    "innerHTML": "<div id=\"gist131045874\" class=\"gist\">\n    <div class=\"gist-file\" translate=\"no\" data-color-mode=\"light\" data-light-theme=\"light\">\n      <div class=\"gist-data\">\n        <div class=\"js-gist-file-update-container js-task-list-container\">\n  <div id=\"file-asdf-tf\" class=\"file my-2\">\n    \n    <div itemprop=\"text\" class=\"Box-body p-0 blob-wrapper data type-hcl  \">\n\n        \n<div class=\"js-check-bidi js-blob-code-container blob-code-content\">\n\n  <template class=\"js-file-alert-template\">\n  <div data-view-component=\"true\" class=\"flash flash-warn flash-full d-flex flex-items-center\">\n  <svg aria-hidden=\"true\" height=\"16\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" data-view-component=\"true\" class=\"octicon octicon-alert\">\n    <path d=\"M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z\"></path>\n</svg>\n    <span>\n      This file contains bidirectional Unicode text that may be interpreted or compiled differently than what appears below. To review, open the file in an editor that reveals hidden Unicode characters.\n      <a class=\"Link--inTextBlock\" href=\"https://github.co/hiddenchars\" target=\"_blank\">Learn more about bidirectional Unicode characters</a>\n    </span>\n\n\n  <div data-view-component=\"true\" class=\"flash-action\">        <a href=\"{{ revealButtonHref }}\" data-view-component=\"true\" class=\"btn-sm btn\">    Show hidden characters\n</a>\n</div>\n</div></template>\n<template class=\"js-line-alert-template\">\n  <span aria-label=\"This line has hidden Unicode characters\" data-view-component=\"true\" class=\"line-alert tooltipped tooltipped-e\">\n    <svg aria-hidden=\"true\" height=\"16\" viewBox=\"0 0 16 16\" version=\"1.1\" width=\"16\" data-view-component=\"true\" class=\"octicon octicon-alert\">\n    <path d=\"M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z\"></path>\n</svg>\n</span></template>\n\n  <table data-hpc class=\"highlight tab-size js-file-line-container js-code-nav-container js-tagsearch-file\" data-tab-size=\"8\" data-paste-markdown-skip data-tagsearch-lang=\"HCL\" data-tagsearch-path=\"asdf.tf\">\n        <tr>\n          <td id=\"file-asdf-tf-L1\" class=\"blob-num js-line-number js-code-nav-line-number js-blob-rnum\" data-line-number=\"1\"></td>\n          <td id=\"file-asdf-tf-LC1\" class=\"blob-code blob-code-inner js-file-line\">asdfasd</td>\n        </tr>\n  </table>\n</div>\n\n\n    </div>\n\n  </div>\n</div>\n\n      </div>\n      <div class=\"gist-meta\">\n        <a href=\"https://gist.github.com/KyMidd/fb0cd10efd25909a66e7ff2016327d69/raw/4e956328ce4aa5e31c4ad5c0303903f84f340336/asdf.tf\" style=\"float:right\" class=\"Link--inTextBlock\">view raw</a>\n        <a href=\"https://gist.github.com/KyMidd/fb0cd10efd25909a66e7ff2016327d69#file-asdf-tf\" class=\"Link--inTextBlock\">\n          asdf.tf\n        </a>\n        hosted with &#10084; by <a class=\"Link--inTextBlock\" href=\"https://github.com\">GitHub</a>\n      </div>\n    </div>\n</div>\n",
    "stylesheet": "https://github.githubassets.com/assets/gist-embed-c38b724e3032.css"
}
```

The template tags are still there in the the `innerHTML` field, thus everything is working fine until that point.

That HTML is then rendered by substack inside a [ProseMirror](https://prosemirror.net/) component, there is likely where the bug happens, since it turns the HTML into:

```html
<link rel="stylesheet" href="https://github.githubassets.com/assets/gist-embed-c38b724e3032.css">
<div id="gist131057207" class="gist">
  <div class="gist-file" data-color-mode="light" data-light-theme="light">
    <div class="gist-data">
      <div class="js-gist-file-update-container js-task-list-container">
        <div id="file-test-txt" class="file my-2">

          <div itemprop="text" class="Box-body p-0 blob-wrapper data type-text  ">


            <div class="js-check-bidi js-blob-code-container blob-code-content">


              <div data-view-component="true" class="flash flash-warn flash-full d-flex flex-items-center">



                <span>
                  This file contains bidirectional Unicode text that may be interpreted or compiled differently than
                  what appears below. To review, open the file in an editor that reveals hidden Unicode characters.
                  <a class="Link--inTextBlock" href="https://github.co/hiddenchars" target="_blank">Learn more about
                    bidirectional Unicode characters</a>
                </span>


                <div data-view-component="true" class="flash-action"> <a href="{{ revealButtonHref }}"
                    data-view-component="true" class="btn-sm btn"> Show hidden characters
                  </a>
                </div>
              </div>

              <span data-view-component="true" class="line-alert tooltipped tooltipped-e">



              </span>

              <table data-hpc=""
                class="highlight tab-size js-file-line-container js-code-nav-container js-tagsearch-file"
                data-tab-size="8" data-paste-markdown-skip="" data-tagsearch-lang="Text" data-tagsearch-path="test.txt">
                <tbody>
                  <tr>
                    <td id="file-test-txt-L1" class="blob-num js-line-number js-code-nav-line-number js-blob-rnum"
                      data-line-number="1"></td>
                    <td id="file-test-txt-LC1" class="blob-code blob-code-inner js-file-line">test</td>
                  </tr>
                </tbody>
              </table>
            </div>


          </div>

        </div>
      </div>

    </div>
    <div class="gist-meta">
      <a href="https://gist.github.com/Maluen/b653862672f998d55bb34a9cb67a3ec3/raw/5ad060ce56a1cf704f9bbc1ed674eccae155f64e/test.txt"
        style="float:right" class="Link--inTextBlock">view raw</a>
      <a href="https://gist.github.com/Maluen/b653862672f998d55bb34a9cb67a3ec3#file-test-txt" class="Link--inTextBlock">
        test.txt
      </a>
      hosted with ❤ by <a class="Link--inTextBlock" href="https://github.com">GitHub</a>
    </div>
  </div>
</div>
```

That is the `template` elements disappear except their content is kept, causing it to be rendered.

**Comment by Maluen** [comment]
They have a custom `GitgistToDOM` component that sanitizes that HTML, only keeping allowed tags and attributes.

![image](https://github.com/Maluen/code-medium/assets/1434239/6bc41bfb-edb6-47f6-b498-ccfce8fc4eb1)

![image](https://github.com/Maluen/code-medium/assets/1434239/66d4db90-2732-4cf6-bde9-7ee3630f6f93)


Unfortunately `template` isn't in the list of allowed tags:

```js
['address', 'article', 'aside', 'footer', 'header', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hgroup', 'main', 'nav', 'section', 'blockquote', 'dd', 'div', 'dl', 'dt', 'figcaption', 'figure', 'hr', 'li', 'main', 'ol', 'p', 'pre', 'ul', 'a', 'abbr', 'b', 'bdi', 'bdo', 'br', 'cite', 'code', 'data', 'dfn', 'em', 'i', 'kbd', 'mark', 'q', 'rb', 'rp', 'rt', 'rtc', 'ruby', 's', 'samp', 'small', 'span', 'strong', 'sub', 'sup', 'time', 'u', 'var', 'wbr', 'caption', 'col', 'colgroup', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'tr']
```

**Comment by Maluen** [comment]
There is a way the extension can avoid the error, it's kind of ugly but it works. Since the gist HTML is fetched with a client-side HTTP call, we can patch the call itself to strip the `template` tags and their content from the HTML:

```javascript
var _open = XMLHttpRequest.prototype.open;
window.XMLHttpRequest.prototype.open = function (method, URL) {
    var _onreadystatechange = this.onreadystatechange,
        _this = this;

    _this.onreadystatechange = function () {
        if (_this.readyState === 4 && _this.status === 200 && ~URL.indexOf('api/v1/github/gist')) {
            try {
                var data = JSON.parse(_this.responseText);

                data.innerHTML = data.innerHTML.replace(/<template ([\s\S]*?)>([\s\S]+?)<\/template>/g, ' ');

                // rewrite responseText
                Object.defineProperty(_this, 'responseText', {
                  value: JSON.stringify(data),
                  configurable: true,
                  enumerable: true,
                });
            } catch (e) {}
        }
        // call original callback
        if (_onreadystatechange) _onreadystatechange.apply(this, arguments);
    };

    // detect any onreadystatechange changing
    Object.defineProperty(this, "onreadystatechange", {
        get: function () {
            return _onreadystatechange;
        },
        set: function (value) {
            _onreadystatechange = value;
        }
    });

    return _open.apply(_this, arguments);
};
```

See https://stackoverflow.com/a/51594799

If we execute this code in the Substack editor on load, all subsequent gist additions will show correctly, including in the published post!

However the following code must be injected in the page itself, executing it in a content script isn't enough given it's a separate context.

**Comment by clintval** [comment]
@Maluen thank you so much for this plugin! I've gotten it to work well on Medium and notice that Substack support may exist. I am trying to avoid the annoying Unicode warning with Substack's builtin functionality.

I installed your plugin from the Chrome webstore today and I can't seem to find the equivalent button when in the Substack editor to add a gist. Do you have a screenshot or pointers on how to get this to work with Substack?

Thanks again for the great tool.

**Comment by Maluen** [comment]
@clintval glad you like it!

The version with "substack support" hasn't been published to the chrome & firefox webstores, it's only available from source. That being said I can look at publishing it if there is interest.

However currently it doesn't extend the Substack editor UI in any way, it only prevents the unicode warning with Substack's builtin functionality.

**Comment by Maluen** [comment]
Actually nevermind, so much time has passed that I forgot I actually also extended the Substack UI with a button to add the gist, even though it's still in the source only. LOL

I'll test this to see if it's still working and publish it to the store if it makes sense.

I'll also have to update the README and webstore page to illustrate the new features, thus it might take a while.

**Comment by clintval** [comment]
Thanks 🙏 @Maluen! I was scanning the source code of this repo and it _**felt**_ like support was already bundled to post-process away that annoying Unicode warning. If it's a small amount of work to actually make it live, I would be greatly appreciative. LMK if you want me to test anything out in advance!

**Comment by Maluen** [comment]
Preventing the unicode warnings works ok both Chrome and Firefox (v128).

The "Insert Github Gist" custom button also seems to still work fine on Chrome:

![Image](https://github.com/user-attachments/assets/b9bfe5cc-5a71-4aa3-b107-8d63998752c0)

On "Firefox" however it doesn't really work, as simulating a Paste event is required to insert the gist URL, but Firefox is not allowing programmatic paste. It just silently fails.

As such currently the button can't be added to Firefox.

**Comment by Maluen** [comment]
I've published version v0.7.0 to the chrome store. It might take few hours to few days to be rolled out automatically.

I'm not going to publish to Firefox store for now given  the lack of the Substack custom button there.

**Comment by fulcrum-blog** [comment]
@Maluen 0.7.0 is live and after I installed it, it works!! Many thanks.

**Comment by clintval** [comment]
@Maluen I installed 0.7.0 (Chrome) and confirm it works, many thanks!!

**Comment by fulcrum-blog** [comment]
Hey @Maluen, thank-you for this plugin, it's extremely useful.  I tried using it today with substack and unfortunately the "bidirectional warning text" issue is coming up for me.  I asked @clintval to give it a go, and he was also to reproduce my issue.  Thank-you for any help you can offer in advance, I really appreciate it!

**Comment by nh13** [comment]
Hey @Maluen, thank-you for this plugin, it's extremely useful.  I tried using it today with substack and unfortunately the "bidirectional warning text" issue is coming up for me.  I asked @clintval to give it a go, and he was also to reproduce my issue.  Thank-you for any help you can offer in advance, I really appreciate it!

**Comment by Maluen** [comment]
@nh13 I just tested with Chrome, using Code Medium 0.7.0 (latest) from the chrome webstore.
I edited a Substack post and added a gist with Code Medium.

Gist is rendering correctly without warnings

![Image](https://github.com/user-attachments/assets/3c9a7ab6-75c3-4170-b50f-9f5991d8e91e)

If you have existing gists that you've previously added before having installed Code Medium, then their warnings won't go away, as their stored HTML already contains the warning. In that case you can just re-add the gist.

**Comment by fulcrum-blog** [comment]
Ah, could it be because we have a custom URL (e.g `https://blog.fulcrumgenomics.com/`)?  Looking at the extension settings, I don't see a way to add to the `Site access`.  Is that something that could be supported?

![Image](https://github.com/user-attachments/assets/9e86ee2e-f214-4dbc-ad76-4c386151e008)

**Comment by Maluen** [comment]
>Ah, could it be because we have a custom URL (e.g https://blog.fulcrumgenomics.com/)? 

Yes, that's exactly it. The extension content script only runs on the URLs listed in the manifest.json, in this case just *.substack.com.

In principle the "run on all urls" rule is supported, but it's very discouraged security and privacy wise, as it's way too broad. It also makes the extension review process more bothersome, thus I'm not going to use it.

Your best bet is to either use it on a substack.com subdomain, or to build the extension directly from source, after having added your URL to:
- https://github.com/Maluen/code-medium/blob/master/src/manifest.js
- https://github.com/Maluen/code-medium/blob/f931b070b33f74975fbef9ab68ac969b5b17bd82/src/content/services/index.js#L7

**Comment by nh13** [comment]
Thank-you for your help, and I totally appreciate it since you're volunteering your time!

Here's the diff (I also ran `yarn update`):

```
diff --git a/src/manifest.js b/src/manifest.js
index 6ce46d7..a09f27d 100644
--- a/src/manifest.js
+++ b/src/manifest.js
@@ -4,6 +4,8 @@ module.exports = (browser, version) => {
   const CHROME = browser === 'chrome';
   const MANIFESTV3 = CHROME;
   const HOST_PERMISSIONS = [
+    "https://blog.fulcrumgenomics.com/**",
+    "https://blog.fulcrumgenomics.com/*",
     "https://github.com/*",
     "https://gist.github.com/*",
     "https://*.substack.com/**", // for InjectService
diff --git a/yarn.lock b/yarn.lock
index df62517..60a05e6 100644
```

Here's screenshots of the plugin showing that it can access the site data:

![Image](https://github.com/user-attachments/assets/9fc52aa6-a1cb-4699-9b0d-342d0ad2337a)

![Image](https://github.com/user-attachments/assets/6db235f0-3ea5-42ea-bca5-872ee146134f)

However, I still get the pre-amble when pasting in a gist (e.g. https://gist.github.com/fulcrum-blog/50cb3ec6795bf9dedf30ad1e95e90d66)

![Image](https://github.com/user-attachments/assets/66ec5f18-fcb8-44a7-a283-7afaadf90465)

**Comment by Maluen** [comment]
@nh13 Did you also add your url to [code-medium/src/content/services/index.js](https://github.com/Maluen/code-medium/blob/f931b070b33f74975fbef9ab68ac969b5b17bd82/src/content/services/index.js#L7) ? Like this

```javascript
 const contentIsSubstack = /(^|\.)substack.com$/i.test(window.location.hostname) || window.location.hostname === 'blog.fulcrumgenomics.com';
```

No need to run `yarn update`, just recompile after doing so and reload the extension.

**Comment by nh13** [comment]
Doh, I missed that.  I added a few more updates too, and now it works.  Thank-you thank-you!

```
diff --git a/src/content/services/index.js b/src/content/services/index.js
index f49a9b1..355c01b 100644
--- a/src/content/services/index.js
+++ b/src/content/services/index.js
@@ -4,7 +4,8 @@ import AppService from './AppService';
 import MediumService from './MediumService';
 import SubstackService from './SubstackService';
 
-const contentIsSubstack = /(^|\.)substack.com$/i.test(window.location.hostname);
+const contentIsSubstack = /(^|\.)substack.com$/i.test(window.location.hostname) || window.location.hostname === 'blog.fulcrumgenomics.com';
+
 
 const { services, start } = createServices([
   ['rpc', ContentRPCService],
diff --git a/src/manifest.js b/src/manifest.js
index 6ce46d7..f5d9040 100644
--- a/src/manifest.js
+++ b/src/manifest.js
@@ -4,6 +4,8 @@ module.exports = (browser, version) => {
   const CHROME = browser === 'chrome';
   const MANIFESTV3 = CHROME;
   const HOST_PERMISSIONS = [
+    "https://blog.fulcrumgenomics.com/**",
+    "https://blog.fulcrumgenomics.com/*",
     "https://github.com/*",
     "https://gist.github.com/*",
     "https://*.substack.com/**", // for InjectService
@@ -82,6 +84,18 @@ module.exports = (browser, version) => {
         ],
         "run_at": "document_idle"
       },
+      {
+        "matches": [
+             "https://blog.fulcrumgenomics.com/*",
+        ],
+        "js": [
+          "content/content_substack.js"
+        ],
+        "css": [
+          "content/content_substack.css"
+        ],
+        "run_at": "document_idle"
+      },
     ],
     "web_accessible_resources": [
       ...(MANIFESTV3 ? [
@@ -89,7 +103,8 @@ module.exports = (browser, version) => {
           "resources": ["app/*.*"],
           "matches": [
             "https://medium.com/*",
-            "https://*.substack.com/*"
+            "https://*.substack.com/*",
+                       "https://blog.fulcrumgenomics.com/*"
           ],
         }
       ] : [
```

**Comment by mdproctor** [comment]
I've installed it from the Chrome Store. There are no menu items added. Clicking on the plugin icon in the toolbar doesn't do anything. Was it supposed to? However, I can paste in a Gist URL, and there is no Unicode error - yay. If I double-click the gist, it opens the side window to log in to Gist. For the first few times, it showed a fetch failure (I had checked that OAuth was approved), and then it mysteriously worked. So I can easily update my gist in the same window - yay, great time saver.

**Comment by Maluen** [comment]
Thanks for the feedback @mdproctor 

>There are no menu items added.

Substack has changed their user interface. Fixed with https://github.com/Maluen/code-medium/commit/28817832df2009d17705c6cf02f7c8ddc32e18f6

> For the first few times, it showed a fetch failure (I had checked that OAuth was approved), and then it mysteriously worked.

I've also got it a couple of times, while other times it worked.
Looking at the network requests in devtools, it shows "net::ERR_NAME_NOT_RESOLVED" when requesting 

`https://github.com/login/oauth/access_token?client_id=[CLIENT_ID]0&client_secret=[CLIENT_SECRET]&code=[CODE]&redirect_uri=https://dganoageikmadjocbmklfgaejpkdigbe.chromiumapp.org/`

I'm assuming this is an issue on github part?
To be safe, the code is now url-encoding the `redirect_uri` part.

**Comment by Maluen** [comment]
v0.8.0 released and submitted for review to the Chrome Store.

No firefox publishing for this version as it doesn't support extending the Substack UI in the first place.
